### ISSUE TITLE
Formatting quirk in bash associative array keys

### ISSUE THREAD

**Issue by dsifford** [mvdan/sh#226]
Hi there,

I ran into a small quirk on bash associative array keys that may or may not be worth fixing.

**Example:**
```sh
declare -A my_array=(
  [foo-bar]='blah blah blah'
)
```
gets formatted to this

```sh
declare -A my_array=(
  [foo - bar]='blah blah blah'
)
```

I should also note that this is not an issue if the user quotes the entire key in this situation.

```sh
# this works fine
declare -A my_array=(
  ['foo-bar']='blah blah blah'
)
```

Thanks again! :wave:

**Comment by mvdan** [comment]
Thanks for opening this issue. Please see the caveats section, which covers this scenario a little bit: https://github.com/mvdan/sh#caveats

However, I've been relaxing said restrictions as of recently, e.g. https://github.com/mvdan/sh/issues/223.

But this is a different issue; if someone writes an arithmetic expression, they want it to be formatted properly. And if they write a string, they don't want any formatting (obviously). So how is the static parser supposed to know which one you went for, in there?

The answer, at the moment, is quotes as you can see. If you have an idea on how we could do better without that restriction, I'm open to suggestions :)

**Comment by dsifford** [comment]
@mvdan Does arithmetic expand in array keys without using double parens?

I was under the assumption that, unless you use a `let` statment, the only way you'll get arithmetic expansion in bash is using the form "`$((varname - 5 + 2))`"

**Comment by mvdan** [comment]
@dsifford yes, if the array is indexed, not associative (i.e. a list, not a map).

```
$ x=(a b c d)
$ echo ${x[1+2]}
d
```

**Comment by mvdan** [comment]
Friendly ping @dsifford - I still think that the current approach is the simplest and safest. We could always try a heuristic-based approach, but note that the parser doesn't execute the shell code, thus it cannot always know whether or not an array was associative.

The only other option that comes to mind would be to *never* parse arithmetic expressions, always leaving them as strings in the AST. However, that seems like a nuclear option. And it would also remove a nice feature from the formatter, which is that it nicely formats arithmetic expressions where it can.

**Comment by dsifford** [comment]
Whoops, sorry for the delay. I'd say go with whatever will be easiest to maintain. This particular issue wasn't that big of a deal for me because I was able to catch it and fix it with the quotes, but I think it could have the potential to break code if someone ran it on a large script and didn't notice it.

If I can be naive and give a suggestion without knowing anything about how this particular codebase or AST works, would it be possible to skip only associative array keys in assignments?

So, for example...

```sh
# This is skipped
declare -A foo=(
  [foo-bar]='baz'
)

# This is not
echo "${foo[foo-bar]}"
```

**Comment by mvdan** [comment]
> I think it could have the potential to break code if someone ran it on a large script and didn't notice it.

That's a good point. Not sure how to avoid that without a nuclear solution, though.

> would it be possible to skip only associative array keys in assignments?

This is what I meant by the heuristic approach before. I believe that would make the behavior of the tool confusing and potentially more problematic:

```
# This is skipped
declare -A a1=(
  [foo-bar]=baz
)

# But what about this
declare -A a2
a2[foo-bar]=baz

# Or when there isn't an easy way to track the declaration
f3() {
    a3[foo-bar]=baz
}
```

Since Bash isn't a typed language, we cannot statically know what kind of a variable some identifier is. As it is now, the tool is a bit dumb, but it is nonetheless right most of the time and it is very predictable. I think that's a good compromise, given the dire condition that Bash as a language puts us in.

**Comment by dsifford** [comment]
Good points. No objections. I'd say leave it alone for now then.

Maybe add a note about this specific variant in the caveats section of the readme as a compromise? I'll be honest and say that I hadn't even considered that the first time I read through that section.

Great discussion. Nice work, as usual. :+1:
